native_client/java/DeepSpeech-jni/org_mozilla_deepspeech_DeepSpeech.cpp

```diff
@@ -0,0 +1,184 @@
+
+    jint state = DS_CreateModel(modelPathCStr, static_cast<unsigned int>(nCep),
+                                static_cast<unsigned int>(nContext), alphaBetCStr,
+                                static_cast<unsigned int>(beamWidth),
+                                &ptr);
+    auto *bufferPtr = (jlong *) (env->GetDirectBufferAddress(modelStatePtr));
+
+    bufferPtr[0] = reinterpret_cast<jlong>(ptr);
+
+    if (isModelPathCopy == JNI_TRUE) {
+        env->ReleaseStringUTFChars(modelPath, modelPathCStr);
+    }
+    return state;
+}
+
+void Java_org_mozilla_deepspeech_DeepSpeech_destroyModel(JNIEnv *, jclass, jlong modelPtr) {
+    DS_DestroyModel(reinterpret_cast<ModelState *>(modelPtr));
+}
+
+jint
+Java_org_mozilla_deepspeech_DeepSpeech_enableDecoderWithLM(JNIEnv *env, jclass, jlong modelStatePtr,
+                                                           jstring alphaBetConfigPath,
+                                                           jstring lmPath,
+                                                           jstring triePath, jfloat alpha,
+                                                           jfloat beta) {
+    jboolean isAlphabetStrCopy, isLmPathCopy, isTriePathCopy;
+    auto alphaBetConfigPathCStr = const_cast<char *>(env->GetStringUTFChars(alphaBetConfigPath,
+                                                                            &isAlphabetStrCopy));
+    auto lmPathCStr = const_cast<char *>(env->GetStringUTFChars(lmPath, &isLmPathCopy));
+    auto triePathCStr = const_cast<char *>(env->GetStringUTFChars(triePath, &isTriePathCopy));
+
+    jint status = DS_EnableDecoderWithLM((ModelState *) modelStatePtr, alphaBetConfigPathCStr,
+                                         lmPathCStr, triePathCStr,
+                                         alpha, beta);
+
+    if (isAlphabetStrCopy == JNI_TRUE) {
+        env->ReleaseStringUTFChars(alphaBetConfigPath, alphaBetConfigPathCStr);
+    }
+    if (isLmPathCopy == JNI_TRUE) {
+        env->ReleaseStringUTFChars(lmPath, lmPathCStr);
+    }
+    if (isTriePathCopy == JNI_TRUE) {
+        env->ReleaseStringUTFChars(triePath, triePathCStr);
+    }
+
+    return status;
+}
+
+jstring
+Java_org_mozilla_deepspeech_DeepSpeech_nSpeechToText(JNIEnv *env, jclass, jlong modelStatePtr,
+                                                     jobject audioBuffer, jlong numSamples,
+                                                     jlong sampleRate) {
+    auto *array = (short *) (env->GetDirectBufferAddress(audioBuffer));
+    char *cStr = DS_SpeechToText((ModelState *) modelStatePtr, array,
+                                 static_cast<unsigned int>(numSamples),
+                                 (unsigned int) sampleRate);
+    if (cStr == nullptr) {
+        return nullptr;
+    }
+    jstring str = env->NewStringUTF(cStr);
+    DS_FreeString(cStr);
+    return str;
+}
+
+jstring
+Java_org_mozilla_deepspeech_DeepSpeech_speechToTextUnsafe(JNIEnv *env, jclass, jlong modelStatePtr,
```

Comment:
Why do we have unsafe and safe versions here ?!

Reply:
To increase performance of the Recognition Stream as Java does not support wrapping a directly allocated buffer object over already allocated memory.

Reply:
Are performances really limited by that ?

Reply:
Yes, I have already talked about that two times.